### PR TITLE
Feature/guilduser timeouts

### PR DESCRIPTION
**Pull request by csmir**
Base implementation > This PR adds the new timeout feature to guild users.

### CODE REVIEW

File: src/Discord.Net.Core/Entities/Users/IGuildUser.cs [hunk 1]
```diff
@@ -211,5 +222,22 @@ public interface IGuildUser : IUser, IVoiceState
...
         Task RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options = null);
+        /// <summary>
+        ///     Adds a timeout based on provided <see cref="TimeSpan"/> to this user in the guild.
+        /// </summary>
+        /// <param name="span">The <see cref="TimeSpan"/> indicating how long a user should be timed out for.</param>
+        /// <param name="options">The options to be used when sending the request.</param>
+        /// <returns>
+        ///     A task that represents the asynchronous timeout creation operation.
+        /// </returns>
+        Task AddTimeOutAsync(TimeSpan span, RequestOptions options = null);
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Users/IGuildUser.cs]: Should be `SetTimeoutAsync`, adding seems like your adding more time to a timeout

File: src/Discord.Net.Rest/API/Rest/ModifyGuildMemberParams.cs [hunk 2]
```diff
@@ -15,5 +16,7 @@ internal class ModifyGuildMemberParams
         public Optional<ulong[]> RoleIds { get; set; }
         [JsonProperty("channel_id")]
         public Optional<ulong?> ChannelId { get; set; }
+        [JsonProperty("communication_disabled_until")]
+        public Optional<DateTimeOffset?> TimeOut { get; set; }
```
Comment (quinchs) [inline src/Discord.Net.Rest/API/Rest/ModifyGuildMemberParams.cs]: Should be `TimeOutUntil`

File: src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs [hunk 3]
```diff
@@ -46,6 +47,8 @@ public int Hierarchy
                 return orderedRoles.Where(x => RoleIds.Contains(x.Id)).Max(x => x.Position);
             }
         }
+        /// <inheritdoc />
+        public DateTimeOffset? TimeOutUntil => DateTimeUtils.FromTicks(_timeOutUntilTicks);
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs]: Maybe `TimeoutDuration` makes more sense?

File: src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs [hunk 4]
```diff
@@ -152,6 +157,12 @@ public Task RemoveRolesAsync(IEnumerable<ulong> roleIds, RequestOptions options
         /// <inheritdoc />
         public Task RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options = null)
             => RemoveRolesAsync(roles.Select(x => x.Id));
+        /// <inheritdoc />
+        public Task AddTimeOutAsync(TimeSpan span, RequestOptions options = null)
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs]: `SetTimeoutAsync` makes more sense

File: src/Discord.Net.Rest/Entities/Users/RestWebhookUser.cs [hunk 5]
```diff
@@ -97,6 +99,12 @@ IGuild IGuildUser.Guild
         /// <inheritdoc />
         Task IGuildUser.RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options) =>
             throw new NotSupportedException("Roles are not supported on webhook users.");
+        /// <inheritdoc />
+        Task IGuildUser.AddTimeOutAsync(TimeSpan span, RequestOptions options) =>
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestWebhookUser.cs]: Should be `SetTimeoutAsync`

File: src/Discord.Net.Rest/Entities/Users/UserHelper.cs [hunk 6]
```diff
@@ -84,5 +88,25 @@ public static async Task RemoveRolesAsync(IGuildUser user, BaseDiscordClient cli
             foreach (var roleId in roleIds)
                 await client.ApiClient.RemoveRoleAsync(user.Guild.Id, user.Id, roleId, options).ConfigureAwait(false);
         }
+
+        public static async Task AddTimeOutAsync(IGuildUser user, BaseDiscordClient client, TimeSpan span, RequestOptions options)
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/UserHelper.cs]: Should be `SetTimeoutAsync`

File: src/Discord.Net.Rest/Entities/Users/UserHelper.cs [hunk 7]
```diff
@@ -84,5 +88,25 @@ public static async Task RemoveRolesAsync(IGuildUser user, BaseDiscordClient cli
...
+        {
+            var apiArgs = new API.Rest.ModifyGuildMemberParams()
+            {
+                TimeOut = new DateTimeOffset(DateTime.UtcNow, span)
+            };
+            if (span.TotalDays >= 28)
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/UserHelper.cs]: Move this check above the apiArgs

File: src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs [hunk 8]
```diff
@@ -90,6 +91,8 @@ public IReadOnlyCollection<SocketRole> Roles
         /// <inheritdoc />
         public DateTimeOffset? PremiumSince => DateTimeUtils.FromTicks(_premiumSinceTicks);
 
+        public DateTimeOffset? TimeOutUntil => DateTimeUtils.FromTicks(_timeOutUntilTicks);
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs]: Should be `TimeoutDuration`

File: src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs [hunk 9]
```diff
@@ -221,6 +226,12 @@ public Task RemoveRolesAsync(IEnumerable<ulong> roleIds, RequestOptions options
         /// <inheritdoc />
         public Task RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options = null)
             => RemoveRolesAsync(roles.Select(x => x.Id));
+        /// <inheritdoc />
+        public Task AddTimeOutAsync(TimeSpan span, RequestOptions options = null)
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs]: Should be `SetTimeoutAsync`

File: src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs [hunk 10]
```diff
@@ -39,6 +39,10 @@ public string Nickname
         public DateTimeOffset? PremiumSince
             => GuildUser.PremiumSince;
 
+        /// <inheritdoc/>
+        public DateTimeOffset? TimeOutUntil
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs]: Should be `TimeoutDuration`

File: src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs [hunk 11]
```diff
@@ -182,6 +186,12 @@ internal void Update(Model model)
         /// <inheritdoc/>
         public Task RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options = null) => GuildUser.RemoveRolesAsync(roles, options);
 
+        /// <inheritdoc/>
+        public Task AddTimeOutAsync(TimeSpan span, RequestOptions options = null) => GuildUser.AddTimeOutAsync(span, options);
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs]: Should be `SetTimeoutAsync`

File: src/Discord.Net.WebSocket/Entities/Users/SocketWebhookUser.cs [hunk 12]
```diff
@@ -129,6 +131,14 @@ internal static SocketWebhookUser Create(SocketGuild guild, ClientState state, M
         /// <exception cref="NotSupportedException">Roles are not supported on webhook users.</exception>
         Task IGuildUser.RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options) =>
             throw new NotSupportedException("Roles are not supported on webhook users.");
+        /// <inheritdoc />
+        /// <exception cref="NotSupportedException">Timeouts are not supported on webhook users.</exception>
+        Task IGuildUser.AddTimeOutAsync(TimeSpan span, RequestOptions options) =>
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketWebhookUser.cs]: Should be `SetTimeoutAsync`

**Review by quinchs — CHANGES_REQUESTED** [review]

File: src/Discord.Net.Rest/Entities/Users/UserHelper.cs [hunk 13]
```diff
@@ -84,5 +88,25 @@ public static async Task RemoveRolesAsync(IGuildUser user, BaseDiscordClient cli
             foreach (var roleId in roleIds)
                 await client.ApiClient.RemoveRoleAsync(user.Guild.Id, user.Id, roleId, options).ConfigureAwait(false);
         }
+
+        public static async Task SetTimeoutAsync(IGuildUser user, BaseDiscordClient client, TimeSpan span, RequestOptions options)
+        {
+            if (span.TotalDays >= 28)
+                throw new ArgumentOutOfRangeException(nameof(span), "Offset cannot be more than 28 days from the current date.");
```
Comment (JerzyKruszewski) [inline src/Discord.Net.Rest/Entities/Users/UserHelper.cs:96]: Just out of curiosity: Why is this feature limited? Is it a mandatory limit by Discord API?

Comment (exsersewo) [reply src/Discord.Net.Rest/Entities/Users/UserHelper.cs:96]: Yeah it's a limit imposed by discord as per https://discord.com/developers/docs/resources/guild#modify-guild-member

image

File: src/Discord.Net.Core/Entities/Users/IGuildUser.cs [hunk 14]
```diff
@@ -85,6 +85,17 @@ public interface IGuildUser : IUser, IVoiceState
...
+        /// <summary>
+        ///     Gets the date and time that indicates if and for how long a user has been timed out.
+        /// </summary>
+        /// <remarks>
+        ///     <see cref="null"/> or a timestamp in the past if the user is not timed out.
+        /// </remarks>
+        /// <returns>
+        ///     A <see cref="DateTimeOffset"/> indicating how long the user will be timed out for.
+        /// </returns>
+        DateTimeOffset? TimeoutDuration { get; }
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Users/IGuildUser.cs]: If the timestamp can be in the past, maybe we should add a getter here that checks if its in the past and returns null.

File: src/Discord.Net.Core/Entities/Users/IGuildUser.cs [hunk 15]
```diff
@@ -211,5 +222,22 @@ public interface IGuildUser : IUser, IVoiceState
         ///     A task that represents the asynchronous role removal operation.
         /// </returns>
         Task RemoveRolesAsync(IEnumerable<IRole> roles, RequestOptions options = null);
+        /// <summary>
+        ///     Adds a timeout based on provided <see cref="TimeSpan"/> to this user in the guild.
```
Comment (quinchs) [inline src/Discord.Net.Core/Entities/Users/IGuildUser.cs]:
```suggestion
        ///     Sets a timeout based on provided <see cref="TimeSpan"/> to this user in the guild.
```

File: src/Discord.Net.Core/Entities/Users/IGuildUser.cs (hunk 14 above)
Comment (quinchs) [reply src/Discord.Net.Core/Entities/Users/IGuildUser.cs]: Ontop of that, I didn't read the correctly yesterday. The name `TimedOutUntil` makes sense

File: src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs [hunk 16]
```diff
@@ -46,6 +47,8 @@ public int Hierarchy
                 return orderedRoles.Where(x => RoleIds.Contains(x.Id)).Max(x => x.Position);
             }
         }
+        /// <inheritdoc />
+        public DateTimeOffset? TimeoutDuration => DateTimeUtils.FromTicks(_timeoutDurationTicks);
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs]:
```suggestion
        public DateTimeOffset? TimedOutUntil => DateTimeUtils.FromTicks(_timeoutDurationTicks);
```

File: src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs [hunk 17]
```diff
@@ -16,6 +16,7 @@ public class RestGuildUser : RestUser, IGuildUser
     {
         #region RestGuildUser
         private long? _premiumSinceTicks;
+        private long? _timeoutDurationTicks;
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs]:
```suggestion
        private long? _timedoutTicks;
```

File: src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs (hunk 16 above)
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestGuildUser.cs]:
```suggestion
        public DateTimeOffset? TimeoutDuration => DateTimeUtils.FromTicks(_timedoutTicks);
```

File: src/Discord.Net.Rest/Entities/Users/RestWebhookUser.cs [hunk 18]
```diff
@@ -62,6 +62,8 @@ IGuild IGuildUser.Guild
         /// <inheritdoc />
         int IGuildUser.Hierarchy => 0;
         /// <inheritdoc />
+        DateTimeOffset? IGuildUser.TimeoutDuration => null;
```
Comment (quinchs) [inline src/Discord.Net.Rest/Entities/Users/RestWebhookUser.cs]:
```suggestion
        DateTimeOffset? IGuildUser.TimedOutUntil => null;
```

File: src/Discord.Net.Rest/Entities/Users/UserHelper.cs (hunk 13 above)
Comment (quinchs) [reply src/Discord.Net.Rest/Entities/Users/UserHelper.cs:96]: There should be another check for negative timespans

File: src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs [hunk 19]
```diff
@@ -20,6 +20,7 @@ public class SocketGuildUser : SocketUser, IGuildUser
     {
         #region SocketGuildUser
         private long? _premiumSinceTicks;
+        private long? _timeoutDurationTicks;
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs]:
```suggestion
        private long? _timedoutTicks;
```

File: src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs [hunk 20]
```diff
@@ -89,6 +90,8 @@ public IReadOnlyCollection<SocketRole> Roles
         public AudioInStream AudioStream => Guild.GetAudioStream(Id);
         /// <inheritdoc />
         public DateTimeOffset? PremiumSince => DateTimeUtils.FromTicks(_premiumSinceTicks);
+        /// <inheritdoc />
+        public DateTimeOffset? TimeoutDuration => DateTimeUtils.FromTicks(_timeoutDurationTicks);
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketGuildUser.cs]:
```suggestion
        public DateTimeOffset? TimedOutUntil => DateTimeUtils.FromTicks(_timedoutTicks);
```

File: src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs [hunk 21]
```diff
@@ -39,6 +39,10 @@ public string Nickname
         public DateTimeOffset? PremiumSince
             => GuildUser.PremiumSince;
 
+        /// <inheritdoc/>
+        public DateTimeOffset? TimeoutDuration
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs]:
```suggestion
        public DateTimeOffset? TimedOutUntil
```

File: src/Discord.Net.WebSocket/Entities/Users/SocketWebhookUser.cs [hunk 22]
```diff
@@ -72,6 +72,8 @@ internal static SocketWebhookUser Create(SocketGuild guild, ClientState state, M
         /// <inheritdoc />
         DateTimeOffset? IGuildUser.PremiumSince => null;
         /// <inheritdoc />
+        DateTimeOffset? IGuildUser.TimeoutDuration => null;
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/Entities/Users/SocketWebhookUser.cs]:
```suggestion
        DateTimeOffset? IGuildUser.TimedOutUntil => null;
```

**Review by quinchs — CHANGES_REQUESTED** [review]

**Review by quinchs — APPROVED** [review]

**Pull request merged**